pilot/pkg/server/instance.go

```diff
@@ -0,0 +1,155 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"sync"
+
+	"istio.io/pkg/log"
+)
+
+type Component func(stop <-chan struct{}) error
+
+// Instance is a server that is composed a number of Component tasks.
+type Instance interface {
+	// Start this Server. Any components that were already added
+	// will be run immediately. If any error is returned,
+	// Start will terminate and return the error immediately.
+	//
+	// Once all startup components have been run, starts a polling
+	// loop to continue monitoring for new components and returns nil.
+	Start(stop <-chan struct{}) error
+
+	// RunComponent adds the given component to the server's run queue.
+	RunComponent(t Component)
+
+	// RunComponentAsync runs the given component asynchronously.
+	RunComponentAsync(t Component)
+
+	// RunComponentAsync runs the given component asynchronously. When
+	// the serer Instance is shutting down, it will wait for the component
+	// to complete before exiting.
+	// Note: this is best effort; a process can die at any time.
+	RunComponentAsyncAndWait(t Component)
+
+	// Wait for this server Instance to shutdown.
+	Wait()
+}
+
+var _ Instance = &instance{}
+
+// New creates a new server Instance.
+func New() Instance {
+	return &instance{
+		done:       make(chan struct{}),
+		components: make(chan Component, 100), // should be enough?
```

Comment:
if I understand right, this means we have a hard limit of 100 multiclusters? Or really ~80 since there are currently ~20 other components. That seems not desired?

Admittedly, 100+ clusters is not common but I don't think we want a hard limit introduced?

Reply:
It's actually simply the number of elements that can be queued for processing at the same time. The server pops them off as it runs them. So I suspect 100 is probably fine, but I'm happy to bump it up.

Reply:
Bumped up to 1000 ... should be more than enough.

Reply:
Don't we not start processing them until start is called, which is after we enqueue all the multicluster parts thought?

Reply:
Yes, but only this (local) cluster is processed during `Start`. The others come in dynamically via remote secrets.